packages/auth/lib/Users.js

```diff
@@ -276,6 +283,38 @@ const denySession = async ({ pgdb, token, email: emailFromQuery, me }) => {
 }
 
 const authorizeSession = async ({ pgdb, tokens, email: emailFromQuery, signInHooks = [], consents = [], requiredFields = [], req, me }) => {
+  const sessions = await pgdb.query(
+    `SELECT DISTINCT s.*
+    FROM
+      sessions s
+    LEFT JOIN
+      tokens t
+      ON t."sessionId" = s.id
+    WHERE
+      t.email = :email AND
+      t."expiresAt" >= now()`,
+    { email: emailFromQuery }
+  )
+
+  await Promise.map(sessions, async session => {
+    await pgdb.public.sessions.updateOne(
+      { id: session.id },
+      {
+        sess: {
+          ...session.sess,
+          authorizeAttempts: ++session.sess.authorizeAttempts || 1
+        }
+      }
+    )
+
+    if (session.sess.authorizeAttempts > MAX_AUTHORIZE_ATTEMPTS) {
+      throw new AuthorizationRateLimitError({
+        email: emailFromQuery,
+        authorizeAttempts: session.sess.authorizeAttempts
+      })
+    }
+  })
+
```

Comment:
querying sessions and updating them should go into a transaction. As it's very possible that there is another simultanious authorizeSession running that is doing the same, or one that is currently authorizing the session. Like this they would overwrite each other without anybody noticing.
Not being able to update a session without querying it first (due deep JSON update) is the root of this problem, which I also don't see a general solution to right now.
I would go for a transaction and `select for update`.